### PR TITLE
Add `logical_right_shift` operation

### PR DESCRIPTION
**Pull request by rdspring1**
Fixes https://github.com/NVIDIA/Fuser/issues/361

### CODE REVIEW

**Comment by rdspring1** [comment]
!build

**Comment by kevinstephano** [comment]
@zasdfgbnm  I just wanted to make sure you think the `logical_right_shift` addition in `arith.h` and `arith.cpp` is consistent with the other shifts you added.

File: csrc/ops/arith.cpp [hunk 1]
```diff
@@ -895,6 +895,47 @@ NVFUSER_DEFINE_INT_ONLY_OP(bitwise_right_shift, Rshift)
...
+    std::is_same<LHS, TensorView*>::value ||
+        std::is_same<RHS, TensorView*>::value,
+    TensorView*,
+    Val*>::type
+logical_right_shift_helper(LHS x, RHS shift) {
+  auto number_of_bits = (x->dtype() == PrimDataType::Int) ? 64 : 32;
+  auto one = IrBuilder::create<Int>(x->container(), 1);
+  auto two = IrBuilder::create<Int>(x->container(), 2);
+  auto thirty_two = IrBuilder::create<Int>(x->container(), number_of_bits);
+  auto mask = sub(pow(two, shift), one);
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:917]: What if `shift` is `64`?

File: csrc/ops/arith.cpp [hunk 2]
```diff
@@ -895,6 +895,47 @@ NVFUSER_DEFINE_INT_ONLY_OP(bitwise_right_shift, Rshift)
...
+TORCH_CUDA_CU_API typename std::conditional<
+    std::is_same<LHS, TensorView*>::value ||
+        std::is_same<RHS, TensorView*>::value,
+    TensorView*,
+    Val*>::type
+logical_right_shift_helper(LHS x, RHS shift) {
+  auto number_of_bits = (x->dtype() == PrimDataType::Int) ? 64 : 32;
+  auto one = IrBuilder::create<Int>(x->container(), 1);
+  auto two = IrBuilder::create<Int>(x->container(), 2);
+  auto thirty_two = IrBuilder::create<Int>(x->container(), number_of_bits);
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:916]: With the recent refactor https://github.com/NVIDIA/Fuser/pull/529, this now should be:
```suggestion
  auto number_of_bits = (x->dtype() == PrimDataType::Int) ? 64L : 32L;
  auto one = IrBuilder::create<Scalar>(x->container(), 1L);
  auto two = IrBuilder::create<Scalar>(x->container(), 2L);
  auto thirty_two = IrBuilder::create<Scalar>(x->container(), number_of_bits);
```

File: csrc/ops/arith.cpp (hunk 1 above)
Comment (rdspring1) [reply csrc/ops/arith.cpp]: I used `-1` to represent the mask when the shift is `sizeof(dtype)`

**Comment by rdspring1** [comment]
!build

File: python_tests/test_python_frontend.py [hunk 3]
```diff
@@ -2241,6 +2241,98 @@ def fusion_func(fd: FusionDefinition):
...
+                4042311,
+                -98914167,
+                5526216,
+            ],
+            device="cuda",
+        )
+
+        for idx, dtype in enumerate(dtypes):
+            current_input = input.to(dtype)
+            num_bits = 32 if (dtype == torch.int32) else 64
```
Comment (zasdfgbnm) [inline python_tests/test_python_frontend.py:2322]: Should we also test shifting `33` or `65` bits? This seems valid in C++
```C++
    std::cout << (uint64_t(1) >> 65) << std::endl;
    std::cout << (uint64_t(-1) >> 65) << std::endl;
    std::cout << (int64_t(1) >> 65) << std::endl;
    std::cout << (int64_t(-1) >> 65) << std::endl;
```
all outputs 0.

File: csrc/ops/arith.cpp [hunk 4]
```diff
@@ -959,6 +959,52 @@ NVFUSER_DEFINE_INT_ONLY_OP(bitwise_right_shift, Rshift)
...
+  auto sizeof_int_dtype = (x->dtype() == PrimDataType::Int) ? 64L : 32L;
+
+  auto neg_one = IrBuilder::create<Val>(x->container(), -1L);
+  auto one = IrBuilder::create<Val>(x->container(), 1L);
+  auto two = IrBuilder::create<Val>(x->container(), 2L);
+  auto num_bits_scalar =
+      IrBuilder::create<Val>(x->container(), sizeof_int_dtype);
+
+  auto mask =
+      where(eq(shift, num_bits_scalar), neg_one, sub(pow(two, shift), one));
```
Comment (zasdfgbnm) [inline csrc/ops/arith.cpp:986]: Should this be the following?
```suggestion
      where(ge(shift, num_bits_scalar), neg_one, sub(pow(two, shift), one));
```

File: python_tests/test_python_frontend.py (hunk 3 above)
Comment (zasdfgbnm) [reply python_tests/test_python_frontend.py:2322]: Well, I don't know. Seems that this is undefined behavior. https://wiki.sei.cmu.edu/confluence/display/c/INT34-C.+Do+not+shift+an+expression+by+a+negative+number+of+bits+or+by+greater+than+or+equal+to+the+number+of+bits+that+exist+in+the+operand
Probably this PR is good as is.

**Review by zasdfgbnm — APPROVED** [review]

**Comment by rdspring1** [comment]
!build

**Pull request merged**